src/pu-command.c

```diff
@@ -0,0 +1,339 @@
+        g_string_append_printf(entry_string, "%*s%s\n",
+                               (gint) (max_len - pu_utf8_strlen(entry_string->str)), "",
+                               entry->description ? entry->description : "");
+        g_string_append(string, g_string_free_and_steal(entry_string));
+    }
+
+    g_string_append_printf(string,
+                           "\nExecute any command with the '-h, --help' option "
+                           "to show command specific help.");
+
+    return g_string_free_and_steal(string);
+}
+
+static gchar *
+get_command_help(PuCommandContext *context,
+                 const gchar *command)
+{
+    GString *string = NULL;
+    PuCommandEntry *entry = NULL;
+
+    g_return_val_if_fail(context != NULL, NULL);
+    g_return_val_if_fail(command != NULL, NULL);
+
+    /* Find command in list of valid command entries */
+    for (gsize i = 0; i < context->entries_len; i++) {
+        if (g_str_equal(command, context->entries[i].name)) {
+            entry = &context->entries[i];
+            break;
+        }
+    }
+
+    if (!entry)
+        return NULL;
+
+    string = g_string_new(NULL);
+    g_string_append(string, entry->description);
+
+    return g_string_free_and_steal(string);
+}
+
+static gchar *
+get_help_text(PuCommandContext *context,
+              const gchar *command)
+{
+    GString *string = NULL;
+
+    g_return_val_if_fail(context != NULL, NULL);
+
+    string = g_string_new(NULL);
+
+    if (command)
+        g_string_append(string, get_command_help(context, command));
+    else
+        g_string_append(string, get_main_help(context));
+
+    g_string_append(string,
+                    "\n\nDocumentation: <https://phytec.github.io/partup>\n"
+                    "Report any issues at: <https://github.com/phytec/partup/issues>");
+
+    return g_string_free_and_steal(string);
+}
+
+void
+pu_command_context_add_entries(PuCommandContext *context,
+                               const PuCommandEntry *entries,
+                               const GOptionEntry *main_option_entries)
+{
+    gsize entries_len;
+
+    g_return_if_fail(context != NULL);
+    g_return_if_fail(entries != NULL);
+
+    if (main_option_entries) {
+        g_option_context_add_main_entries(context->option_context,
+                                          main_option_entries, NULL);
+    } else {
+        GOptionEntry entries[] = { { NULL } };
+        g_option_context_add_main_entries(context->option_context, entries, NULL);
+    }
+
+    for (entries_len = 0; entries[entries_len].name != NULL; entries_len++);
+    g_return_if_fail(entries_len <= G_MAXSIZE - context->entries_len);
+
+    context->entries = g_renew(PuCommandEntry, context->entries,
+                               context->entries_len + entries_len);
+
+    if (entries_len != 0) {
+        memcpy(context->entries + context->entries_len, entries,
+               sizeof(PuCommandEntry) * entries_len);
+    }
+
+    context->entries_len += entries_len;
```

Comment:
I doesn't get this complete, but entries_len seams to be an index and you add it every round. So
`context->entries_len + 0 + 1 + 2 +3` Is this right?

Reply:
`context->entries_len` stores the number of already added command entries. `entries_len` is a local variable storing the number of command entries about to be added with this call. As `pu_command_context_add_entries()` can be called multiple times, we want to increase the amount of command entries each time by the size of the entry array `entries` passed to this function.

Reply:
sorry, I misread the for loop. I now see its a one line "calc the entries" command